File: src/com/fictional_particle_sim/physicals/Particle.java
package com.fictional_particle_sim.physicals;

import com.fictional_particle_sim.geometrics.*;
import com.fictional_particle_sim.geometrics.Point;

import java.awt.*;
import java.util.Objects;

import static com.fictional_particle_sim.Constants.*;

public class Particle {

    public Point pos, lastPos;
    public Vector vel;
    public double mass, charge, maxCharge;
    public boolean fixedCharge, fixedMaxCharge, fixedVel;

    public Color chargeColor, maxChargeColor;

    public Particle(Point pos) {
        this.lastPos = pos;
        this.pos = pos;
        this.vel = new Vector();

        this.mass = 1;
        this.charge = 0;
        this.maxCharge = 0;

        this.fixedCharge = false;
        this.fixedVel = false;
    }
    public Particle(Point pos, Vector vel) {
        this.lastPos = pos;
        this.pos = pos;
        this.vel = vel.center(pos);

        this.mass = 1;
        this.charge = 0;
        this.maxCharge = 0;

        this.fixedCharge = false;
        this.fixedVel = false;
    }
    public Particle(Point pos, double mass, double charge) {
        this.lastPos = pos;
        this.pos = pos;
        this.vel = new Vector();

        this.mass = mass;
        this.charge = charge;
        this.maxCharge = 0;

        this.fixedCharge = false;
        this.fixedVel = false;
    }
    public Particle(Point pos, Vector vel, double mass, double charge) {
        this.lastPos = pos;
        this.pos = pos;
        this.vel = vel.center(pos);

        this.mass = mass;
        this.charge = charge;
        this.maxCharge = 0;

        this.fixedCharge = false;
        this.fixedVel = false;
    }
    public Particle(Point pos, boolean fixedCharge, boolean fixedMaxCharge, boolean fixedVel) {
        this.lastPos = pos;
        this.pos = pos;
        this.vel = new Vector();

        this.mass = 1;
        this.charge = 0;
        this.maxCharge = 0;

        this.fixedCharge = fixedCharge;
        this.fixedVel = fixedVel;
    }
    public Particle(Point pos, Vector vel, boolean fixedCharge, boolean fixedMaxCharge, boolean fixedVel) {
        this.lastPos = pos;
        this.pos = pos;
        this.vel = vel.center(pos);

        this.mass = 1;
        this.charge = 0;
        this.maxCharge = 0;

        this.fixedCharge = fixedCharge;
        this.fixedVel = fixedVel;
    }
    public Particle(Point pos, double mass, double charge, boolean fixedCharge, boolean fixedMaxCharge, boolean fixedVel) {
        this.lastPos = pos;
        this.pos = pos;
        this.vel = new Vector();

        this.mass = mass;
        this.charge = charge;
        this.maxCharge = 0;

        this.fixedCharge = fixedCharge;
        this.fixedVel = fixedVel;
    }
    public Particle(Point pos, Vector vel, double mass, double charge, boolean fixedCharge, boolean fixedMaxCharge, boolean fixedVel) {
        this.lastPos = pos;
        this.pos = pos;
        this.vel = vel.center(pos);

        this.mass = mass;
        this.charge = charge;
        this.maxCharge = 0;

        this.fixedCharge = fixedCharge;
        this.fixedVel = fixedVel;
    }

    //Generates a Particle with Random Position
    public static Particle random() {
        return new Particle(new Point(Math.random() * SCALE_WIDTH, Math.random() * SCALE_HEIGHT), 1.0, (Math.random() * 2) - 1);
    }
    public static Particle random(double mass) {
        return new Particle(new Point(Math.random() * SCALE_WIDTH, Math.random() * SCALE_HEIGHT), mass, (Math.random() * 2) - 1);
    }
    public static Particle random(Vector vel, double mass) {
        return new Particle(new Point(Math.random() * SCALE_WIDTH, Math.random() * SCALE_HEIGHT), vel, mass, (Math.random() * 2) - 1);
    }
    public static Particle random(double mass, double charge) {
        return new Particle(new Point(Math.random() * SCALE_WIDTH, Math.random() * SCALE_HEIGHT), mass, charge);
    }
    public static Particle random(Vector vel, double mass, double charge) {
        return new Particle(new Point(Math.random() * SCALE_WIDTH, Math.random() * SCALE_HEIGHT), vel, mass, charge);
    }
    public static Particle random(double mass, double charge, boolean fixedCharge, boolean fixedMaxCharge, boolean fixedVel) {
        return new Particle(new Point(Math.random() * SCALE_WIDTH, Math.random() * SCALE_HEIGHT), mass, charge, fixedCharge, fixedMaxCharge, fixedVel);
    }
    public static Particle random(Vector vel, double mass, double charge, boolean fixedCharge, boolean fixedMaxCharge, boolean fixedVel) {
        return new Particle(new Point(Math.random() * SCALE_WIDTH, Math.random() * SCALE_HEIGHT), vel, mass, charge, fixedCharge, fixedMaxCharge, fixedVel);
    }

    public void setChargeColor(Color chargeColor) {
        this.chargeColor = chargeColor;
    }

    public void setPos(Point pos) {
        this.pos = pos;
    }
    public void setVel(Vector vel) {
        this.vel = vel;
    }

    public void setMass(double mass) {
        this.mass = mass;
    }
    public void setCharge(double charge) {
        this.charge = charge;
    }
    public void setMaxCharge(double maxCharge) {
        this.maxCharge = maxCharge;
    }

    public void setFixedCharge(boolean fixedCharge) {
        this.fixedCharge = fixedCharge;
    }
    public void setFixedVel(boolean fixedVel) {
        this.fixedVel = fixedVel;
    }

    // Calculations for Individual Particles
    public Vector force(Particle b) {

        double dist = this.pos.dist(b.pos);
        if (dist != 0) {

            Vector direct = new Vector(this.pos, b.pos).norm();
            double distCube = TWO_PI_SQUARED * Math.pow(dist, 3);
            if (dist > 2 * MIN_DIST) {
                double forceM = this.charge * b.charge / distCube;
                return direct.scalar(forceM);
            } else if (dist > MIN_DIST) {
                double forceM = (this.charge * b.charge/2 - PUSHBACK_FORCE * this.vel.proj(direct).sub(b.vel.proj(direct)).magnitude()) / distCube;
                return direct.scalar(forceM);
            } else if (dist == MIN_DIST) {
                return new Vector();
            } else {
                double forceM = PUSHBACK_FORCE * this.vel.proj(direct).sub(b.vel.proj(direct)).magnitude() / distCube;
                return direct.scalar(forceM);
            }

        } else return new Vector();
    }
    public Vector acc(Vector force) {

        return force.scalar( 1 / this.mass );
    }
    public void applyAcc(Vector acc) {

        vel = vel.add(acc);
        if (vel.magnitude() > MAX_VEL) {
            vel = vel.norm().scalar(MAX_VEL);
        }
    }
    public void applyVel(boolean updateLastPos) {

        if (updateLastPos) lastPos = pos;
        pos = pos.add(vel.center().scalar(SPF).end);
    }
    public double maxCharge(Particle b) {

        if (!this.equals(b)) {
            double dist = this.pos.dist(b.pos);

            if (dist != 0) {
                return CHARGE_FORCE * b.charge / (TWO_PI_SQUARED * Math.pow(dist, 3));
            } else return b.charge / Math.abs(b.charge);
        } else return 0;
    }
    public double charge(Particle b) {

        if (!this.equals(b)) {
            double dist = this.pos.dist(b.pos);
            if (dist != 0) {
                return DELTA_CHARGE_FORCE * b.charge / (2 * TWO_PI_SQUARED * Math.pow(dist, 2));
            } else return b.charge / Math.abs(b.charge);
        } else return 0;

    }
    public void collide(Barrier b) {
        switch (b.shape) {
            case "LINE" -> {
                if (b.line.between(this.pos, this.lastPos)) {
                    pos = pos.sub(pos.disp(b.line).center().end.mult(1.001));
                    vel = vel.sub(vel.perp(b.line).scalar(2));
                }
            } case "CIRCLE" -> {
                if (this.pos.dist(b.center) <= b.radius && this.lastPos.dist(b.center) > b.radius) {
                    pos = pos.add(new Vector(b.center, pos).norm().center().scalar(b.radius * 1.001 - pos.dist(b.center)).end);
                    vel = vel.sub(vel.proj(new Vector(pos, b.center)).scalar(2));
                } else if (this.pos.dist(b.center) > b.radius && this.lastPos.dist(b.center) <= b.radius) {
                    pos = pos.sub(new Vector(pos, b.center).norm().center().scalar(b.radius * 0.999 - pos.dist(b.center)).end);
                    vel = vel.sub(vel.proj(new Vector(pos, b.center)).scalar(2));
                }
            } case "RECTANGLE" -> {
                if (this.pos.inside(b.boundingBox()) && !this.lastPos.inside(b.boundingBox())) {
                    if (lastPos.x < b.topLeft.x) {
                        pos = new Point(b.topLeft.x - 0.0001, pos.y);
                        vel = new Vector(new Point(-vel.center().end.x, vel.center().end.y)).shift(pos);
                    } else if (lastPos.x > b.bottomRight.x) {
                        pos = new Point(b.bottomRight.x + 0.0001, pos.y);
                        vel = new Vector(new Point(-vel.center().end.x, vel.center().end.y)).shift(pos);
                    } else if (lastPos.y < b.topLeft.y) {
                        pos = new Point(pos.x, b.topLeft.y - 0.0001);
                        vel = new Vector(new Point(vel.center().end.x,-vel.center().end.y)).shift(pos);
                    } else if (lastPos.y > b.bottomRight.y) {
                        pos = new Point(pos.x, b.bottomRight.y + 0.0001);
                        vel = new Vector(new Point(vel.center().end.x,-vel.center().end.y)).shift(pos);
                    }
                } else if (!this.pos.inside(b.boundingBox()) && this.lastPos.inside(b.boundingBox())) {
                    if (pos.x < b.topLeft.x) {
                        pos = new Point(b.topLeft.x + 0.0001, pos.y);
                        vel = new Vector(new Point(-vel.center().end.x, vel.center().end.y)).shift(pos);
                    } else if (pos.x > b.bottomRight.x) {
                        pos = new Point(b.bottomRight.x - 0.0001, pos.y);
                        vel = new Vector(new Point(-vel.center().end.x, vel.center().end.y)).shift(pos);
                    } else if (pos.y < b.topLeft.y) {
                        pos = new Point(pos.x, b.topLeft.y + 0.0001);
                        vel = new Vector(new Point(vel.center().end.x,-vel.center().end.y)).shift(pos);
                    } else if (pos.y > b.bottomRight.y) {
                        pos = new Point(pos.x, b.bottomRight.y - 0.0001);
                        vel = new Vector(new Point(vel.center().end.x,-vel.center().end.y)).shift(pos);
                    }
                }
            }
        }
    }
    public boolean collide(Field f) {
        return switch (f.shape) {
            case "RECTANGLE" -> pos.inside(f.boundingBox);
            case "CIRCLE" -> pos.dist(f.center) <= f.radius;
            default -> false;
        };
    }
    public void chargeColor() {
        if (charge == 0) {
            chargeColor = new Color(128, 128, 128);
        } else if (charge >= 1) {
            chargeColor = new Color(36, 162, 26);
        } else if (charge <= -1) {
            chargeColor = new Color(229, 43, 43);
        } else if (charge > 0) {
            chargeColor = new Color((int)(36*(charge) + 128*(1-charge)), (int)(162*(charge) + 128*(1-charge)), (int)(26*(charge) + 128*(1-charge)));
        } else {
            chargeColor = new Color((int)(229*(-charge) + 128*(1+charge)), (int)(43*(-charge) + 128*(1+charge)), (int)(43*(-charge) + 128*(1+charge)));
        }
        if(mass < 0) {
            chargeColor = new Color(255 - chargeColor.getRed(), 255 - chargeColor.getGreen(), 255 - chargeColor.getBlue());
        }

        if (Math.abs(maxCharge) > Math.abs(charge)){
            if (maxCharge == 0) {
                maxChargeColor = new Color(128, 128, 128);
            } else if (maxCharge >= 1) {
                maxChargeColor = new Color(36, 162, 26);
            } else if (maxCharge <= -1) {
                maxChargeColor = new Color(229, 43, 43);
            } else if (maxCharge > 0) {
                maxChargeColor = new Color((int) (36 * (maxCharge) + 128 * (1 - maxCharge)), (int) (162 * (maxCharge) + 128 * (1 - maxCharge)), (int) (26 * (maxCharge) + 128 * (1 - maxCharge)));
            } else {
                maxChargeColor = new Color((int) (229 * (-maxCharge) + 128 * (1 + maxCharge)), (int) (43 * (-maxCharge) + 128 * (1 + maxCharge)), (int) (43 * (-maxCharge) + 128 * (1 + maxCharge)));
            }
            if (mass < 0) {
                maxChargeColor = new Color(255 - maxChargeColor.getRed(), 255 - maxChargeColor.getGreen(), 255 - maxChargeColor.getBlue());
            }
        } else maxChargeColor = chargeColor;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Particle particle = (Particle) o;
        return Double.compare(particle.mass, mass) == 0 && Double.compare(particle.charge, charge) == 0 && Double.compare(particle.maxCharge, maxCharge) == 0 && fixedCharge == particle.fixedCharge && fixedVel == particle.fixedVel && Objects.equals(pos, particle.pos) && Objects.equals(vel, particle.vel) && Objects.equals(chargeColor, particle.chargeColor);
    }

    @Override
    public int hashCode() {
        return Objects.hash(pos, vel, mass, charge, maxCharge, fixedCharge, fixedVel, chargeColor);
    }

    @Override
    public String toString() {
        return "Particle{" +
                "pos = " + pos +
                ", lastPos = " + lastPos +
                ", vel = " + vel.center().end +
                ", mass = " + mass +
                ", charge = " + charge +
                ", max charge = " + maxCharge +
                '}';
    }
}

File: src/com/fictional_particle_sim/physicals/Particle.kt
package com.fictional_particle_sim.physicals

import com.fictional_particle_sim.util.Constants
import com.fictional_particle_sim.geometrics.Point
import com.fictional_particle_sim.geometrics.Vector
import com.fictional_particle_sim.physicals.Shape.*
import com.fictional_particle_sim.physicals.Type.*
import com.fictional_particle_sim.util.Constants.Companion.MAX_ATTRACTOR_CHARGE
import com.fictional_particle_sim.util.Constants.Companion.MAX_CHARGE
import java.awt.Color
import kotlin.math.abs
import kotlin.math.pow
import kotlin.math.round
import kotlin.math.sign

class Particle(var pos: Point, var lastPos: Point = pos,
               var vel: Vector = Vector(),
               var mass: Double = 1.0, var charge: Double = 0.0, var targetCharge: Double = 0.0,
               var fixedCharge: Boolean = false, var fixedMaxCharge: Boolean = false, var fixedVel: Boolean = false, var type: Type = CHARGED,
               var chargeColor: Color = Color.GRAY, var targetChargeColor: Color = chargeColor) {


    fun force(b: Particle): Vector {
        val dist = pos.dist(b.pos)
        val tempCharge = when (type){
            UNIVERSAL_ATTRACTOR -> when (b.type) {
                UNIVERSAL_ATTRACTOR -> -MAX_ATTRACTOR_CHARGE
                CHARGED, UNIVERSAL_REPELLER -> MAX_ATTRACTOR_CHARGE
            }
            UNIVERSAL_REPELLER -> when (b.type) {
                UNIVERSAL_ATTRACTOR -> MAX_ATTRACTOR_CHARGE
                CHARGED, UNIVERSAL_REPELLER -> -MAX_ATTRACTOR_CHARGE
            }
            CHARGED -> when (b.type) {
                UNIVERSAL_ATTRACTOR, UNIVERSAL_REPELLER -> 1.0
                CHARGED -> charge
            }
        }
        val tempChargeB = when (b.type) {
            UNIVERSAL_ATTRACTOR, UNIVERSAL_REPELLER -> MAX_ATTRACTOR_CHARGE
            CHARGED -> when (type) {
                UNIVERSAL_ATTRACTOR, UNIVERSAL_REPELLER -> 1.0
                CHARGED -> b.charge
            }
        }
        return if (dist != 0.0) {
            val direct: Vector = Vector(pos, b.pos).norm()
            val distCube = Constants.TWO_PI_SQUARED * dist.pow(3.0)
            when {
                dist > 2 * Constants.MIN_DIST -> {
                    val forceM = tempCharge * tempChargeB / distCube
                    direct * forceM
                }
                dist > Constants.MIN_DIST -> {
                    val forceM: Double = (tempCharge * tempChargeB / 2 - Constants.PUSHBACK_FORCE * (vel.proj(direct) - b.vel.proj(direct)).magnitude()) / distCube
                    direct * forceM
                }
                dist == Constants.MIN_DIST -> {
                    val forceM = tempCharge * tempChargeB / distCube
                    if (forceM > 0) Vector() else direct * forceM
                }
                else -> {
                    var forceM = tempCharge * tempChargeB / distCube
                    if (forceM >= 0) {
                        forceM = Constants.PUSHBACK_FORCE * (vel.proj(direct) - b.vel.proj(direct)).magnitude() / distCube
                    }
                    direct * forceM
                }
            }
        } else Vector()
    }
    fun acc(force: Vector): Vector = force/mass
    fun applyAcc(acc: Vector) {
        vel += acc
        if (vel.magnitude() > Constants.MAX_VEL) {
            vel = vel.norm() * Constants.MAX_VEL
        }
    }
    fun applyVel(updateLastPos: Boolean) {
        if (updateLastPos) lastPos = pos
        pos += (vel.center() * (Constants.SPF)).end
    }
    fun maxCharge(b: Particle): Double {
        return if (this != b && (b.type != UNIVERSAL_ATTRACTOR || b.type != UNIVERSAL_ATTRACTOR)) {
            val dist = pos.dist(b.pos)
            if (dist != 0.0) {
                Constants.CHARGE_FORCE * b.charge / (Constants.TWO_PI_SQUARED * dist.pow(3.0))
            } else sign(b.charge) * MAX_CHARGE
        } else 0.0
    }
    fun charge(b: Particle): Double {
        return if (this != b && (b.type != UNIVERSAL_ATTRACTOR || b.type != UNIVERSAL_ATTRACTOR)) {
            val dist = pos.dist(b.pos)
            if (dist != 0.0) {
                Constants.DELTA_CHARGE_FORCE * b.charge / (2 * Constants.TWO_PI_SQUARED * dist.pow(2.0))
            } else sign(b.charge) * MAX_CHARGE
        } else 0.0
    }
    fun collide(b: Barrier) {
        when (b.shape) {
            LINE -> {
                if (b.line.between(pos, lastPos)) {
                    pos -= (pos.disp(b.line).center().end * (1.001))
                    vel -= (vel.perp(b.line) * (2.0))
                }
            }
            CIRCLE -> {
                if (this.pos.dist(b.center) <= b.radius && lastPos.dist(b.center) > b.radius) {
                    pos += (Vector(b.center, pos).norm().center() * (b.radius * 1.001 - pos.dist(b.center))).end
                    vel -= (vel.proj(Vector(pos, b.center)) * (2.0))
                } else if (this.pos.dist(b.center) > b.radius && lastPos.dist(b.center) <= b.radius) {
                    pos -= (Vector(pos, b.center).norm().center() * (b.radius * 0.999 - pos.dist(b.center))).end
                    vel -= (vel.proj(Vector(pos, b.center)) * (2.0))
                }
            }
            RECTANGLE -> {
                if (pos.inside(b.boundingBox()) && !lastPos.inside(b.boundingBox())) {
                    when {
                        lastPos.x < b.topLeft.x -> {
                            pos = Point(b.topLeft.x - 0.0001, pos.y)
                            vel = Vector(Point(-vel.center().end.x, vel.center().end.y)) + pos
                        }
                        lastPos.x > b.bottomRight.x -> {
                            pos = Point(b.bottomRight.x + 0.0001, pos.y)
                            vel = Vector(Point(-vel.center().end.x, vel.center().end.y)) + pos
                        }
                        lastPos.y < b.topLeft.y -> {
                            pos = Point(pos.x, b.topLeft.y - 0.0001)
                            vel = Vector(Point(vel.center().end.x, -vel.center().end.y)) + pos
                        }
                        lastPos.y > b.bottomRight.y -> {
                            pos = Point(pos.x, b.bottomRight.y + 0.0001)
                            vel = Vector(Point(vel.center().end.x, -vel.center().end.y)) + pos
                        }
                    }
                } else if (!pos.inside(b.boundingBox()) && lastPos.inside(b.boundingBox())) {
                    when {
                        pos.x < b.topLeft.x -> {
                            pos = Point(b.topLeft.x + 0.0001, pos.y)
                            vel = Vector(Point(-vel.center().end.x, vel.center().end.y)) + pos
                        }
                        pos.x > b.bottomRight.x -> {
                            pos = Point(b.bottomRight.x - 0.0001, pos.y)
                            vel = Vector(Point(-vel.center().end.x, vel.center().end.y)) + pos
                        }
                        pos.y < b.topLeft.y -> {
                            pos = Point(pos.x, b.topLeft.y + 0.0001)
                            vel = Vector(Point(vel.center().end.x, -vel.center().end.y)) + pos
                        }
                        pos.y > b.bottomRight.y -> {
                            pos = Point(pos.x, b.bottomRight.y - 0.0001)
                            vel = Vector(Point(vel.center().end.x, -vel.center().end.y)) + pos
                        }
                    }
                }
            }

        }
    }
    fun collide(f: Field): Boolean {
        return when (f.shape) {
            RECTANGLE -> pos.inside(f.boundingBox())
            CIRCLE -> pos.dist(f.center) <= f.radius
            else -> false
        }
    }

    fun chargeColor() {
        val gray = when (type) {
            CHARGED -> Color(128, 128, 128)
            UNIVERSAL_ATTRACTOR -> Color(252, 211, 3)
            UNIVERSAL_REPELLER -> Color(3, 44, 252)
        }
        val red = when (type) {
            CHARGED -> Color(36, 162, 26)
            UNIVERSAL_ATTRACTOR -> Color(252, 211, 3)
            UNIVERSAL_REPELLER -> Color(3, 44, 252)
        }
        val green = when (type) {
            CHARGED -> Color(229, 43, 43)
            UNIVERSAL_ATTRACTOR -> Color(252, 211, 3)
            UNIVERSAL_REPELLER -> Color(3, 44, 252)
        }
        val tempMaxCharge = when (type) {
            UNIVERSAL_ATTRACTOR, UNIVERSAL_REPELLER -> MAX_ATTRACTOR_CHARGE
            CHARGED -> MAX_CHARGE
        }

        chargeColor = when {
            charge == 0.0 -> gray
            charge >= tempMaxCharge -> red
            charge <= -tempMaxCharge -> green
            charge > 0 -> Color((red.red * charge/tempMaxCharge + gray.red * (1 - charge/tempMaxCharge)).toInt(), (red.green * charge/tempMaxCharge + gray.green * (1 - charge/tempMaxCharge)).toInt(), (red.blue * charge/tempMaxCharge + gray.blue * (1 - charge/tempMaxCharge)).toInt())
            else -> Color((green.red * -charge/tempMaxCharge + gray.red * (1 + charge/tempMaxCharge)).toInt(), (green.green * -charge/tempMaxCharge + gray.green * (1 + charge/tempMaxCharge)).toInt(), (green.blue * -charge/tempMaxCharge + gray.blue * (1 + charge/tempMaxCharge)).toInt())
        }
        if (mass < 0) {
            chargeColor = Color(255 - chargeColor.red, 255 - chargeColor.green, 255 - chargeColor.blue)
        }

        if (abs(targetCharge) > abs(charge)) {
            targetChargeColor = when {
                targetCharge == 0.0 -> gray
                targetCharge >= tempMaxCharge -> red
                targetCharge <= -tempMaxCharge -> green
                targetCharge > 0 -> Color((red.red * targetCharge/tempMaxCharge + gray.red * (1 - targetCharge/tempMaxCharge)).toInt(), (red.green * targetCharge/tempMaxCharge + gray.green * (1 - targetCharge/tempMaxCharge)).toInt(), (red.blue * targetCharge/tempMaxCharge + gray.blue * (1 - targetCharge/tempMaxCharge)).toInt())
                else -> Color((green.red * -targetCharge/tempMaxCharge + gray.red * (1 + targetCharge/tempMaxCharge)).toInt(), (green.green * -targetCharge/tempMaxCharge + gray.green * (1 + targetCharge/tempMaxCharge)).toInt(), (green.blue * -targetCharge/tempMaxCharge + gray.blue * (1 + targetCharge/tempMaxCharge)).toInt())
            }
            if (mass < 0) {
                targetChargeColor = Color(255 - targetChargeColor.red, 255 - targetChargeColor.green, 255 - targetChargeColor.blue)
            }
        } else targetChargeColor = chargeColor
    }

    override fun toString(): String {
        return "position: $pos, velocity: ${vel.toString(true)}, Mass:$mass, Charge: ${round(charge * 10000)/10000}, Max Charge: ${round(targetCharge * 10000)/10000}"
    }
}

enum class Type {
    CHARGED, UNIVERSAL_ATTRACTOR, UNIVERSAL_REPELLER
}

